File: usaco/Problem_2_Farmer_John_s_Favorite_Operation.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

void solve() {
    long long n, m;
    cin >> n >> m;
    vector<long long> a(n);
    for(int i = 0; i < n; i++) {
        cin >> a[i];
    }
    
    long long best = LLONG_MAX;
    vector<long long> r(n);
    for(int i = 0; i < n; i++) {
        r[i] = a[i] % m;
    }
    sort(r.begin(), r.end());
    
    vector<long long> p(n + 1, 0);
    for(int i = 0; i < n; i++) {
        p[i + 1] = p[i] + r[i];
    }
    
    for(int i = 0; i < n; i++) {
        long long curr = 0;
        int l = i;
        int rr = n - 1;
        int temp = l;
        
        while(l <= rr) {
            int mid = (l + rr) / 2;
            if(r[mid] - r[i] <= m + r[i] - r[mid]) {
                temp = mid;
                l = mid + 1;
            } else {
                rr = mid - 1;
            }
        }
        
        curr += p[temp + 1] - p[i] - (long long)(temp - i + 1) * r[i];
        curr += (long long)(r[i] + m) * (n - (temp + 1)) - (p[n] - p[temp + 1]);
        
        l = 0;
        rr = i;
        int temp2 = i;
        
        while(l <= rr) {
            int mid = (l + rr) / 2;
            if(r[i] - r[mid] <= m - (r[i] - r[mid])) {
                rr = mid - 1;
                temp2 = mid;
            } else {
                l = mid + 1;
            }
        }
        
        curr += (long long)(i + 1 - temp2) * r[i] - (p[i + 1] - p[temp2]);
        curr += (long long)(m - r[i]) * temp2 + p[temp2];
        best = min(best, curr);
    }
    cout << best << "\n";
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int t;
    cin >> t;
    while(t--) {
        solve();
    }
    return 0;
}

File: usaco/Problem_2_Farmer_John_s_Favorite_Operation.py
import sys
def input(): return sys.stdin.readline().rstrip("\r\n")
inf = float('inf')
def solve():
    n, m = list(map(int, input().split()))
    a = list(map(int, input().split()))
    inf = float('inf')
    best = inf
    r = [x % m for x in a]
    r.sort()
    p = [0]
    for x in r:
        p.append(p[-1] + x)
    for i in range(n):
        curr = 0
        l = i
        rr = n - 1
        temp = l
        while l <= rr:
            mid = (l + rr) // 2
            if r[mid] - r[i] <= m + r[i] - r[mid]:
                temp = mid
                l = mid + 1
            else:
                rr = mid -1
        curr += p[temp + 1] - p[i] - (temp - i + 1) * r[i]
        curr += (r[i] + m) * (n - (temp + 1)) - (p[n] - p[temp + 1])
        l = 0
        rr = i
        temp2 = i
        while l <= rr:
            mid = (l + rr) // 2
            if r[i] - r[mid] <= m - (r[i] - r[mid]):
                rr = mid - 1
                temp2 = mid
            else:
                l = mid + 1
        curr += (i + 1 - temp2) * r[i] - (p[i + 1] - p[temp2])
        curr += (m - r[i]) * temp2 + p[temp2]
        best = min(best, curr)
    print(best)


for _ in range(int(input())):
    solve()
